### PR TITLE
core: stdcm: set block to route map even if path ended

### PR DESCRIPTION
**Pull request by eckter**
Small bugfix, the other loop sometimes stops early

### CODE REVIEW

**Review by Erashin — APPROVED** [review]

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt [hunk 1]
```diff
@@ -273,6 +273,8 @@ private class InfraExplorerImpl(
         var pathStarted = incrementalPath.pathStarted
 
         for (block in routeBlocks) {
+            blockRoutes.getOrPut(block) { mutableSetOf() }.add(route)
```
Comment (shenriotpro) [inline core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt]: Isn't there a risk of ending up in an inconsistent state (also related with the doc of the function)?

Comment (eckter) [reply core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt]: Good catch, in fact there's some general issues in how we handle this map. I've fixed it

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt [hunk 2]
```diff
@@ -271,8 +271,11 @@ private class InfraExplorerImpl(
         var nBlocksToSkip = 0
         val pathFragments = arrayListOf<PathFragment>()
         var pathStarted = incrementalPath.pathStarted
+        val addedBlocks = mutableListOf<BlockId>()
 
         for (block in routeBlocks) {
+            addedBlocks.add(block)
+            blocks.add(block)
```
Comment (shenriotpro) [inline core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt]: Do we want to update the doc and clarify `blocks` may be changed even if false is returned?

Comment (eckter) [reply core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt]: I went too fast when fixing that and only reverted part of the change, the state is now unchanged when false is returned

File: core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt [hunk 3]
```diff
@@ -298,14 +300,14 @@ private class InfraExplorerImpl(
...
-                blockRoutes.getOrPut(block) { mutableSetOf() }.add(route)
                 pathStarted = true
                 if (endPath) break // Can't extend any further
             }
         }
         assert(seenFirstBlock)
-        blocks.addAll(routeBlocks)
+        blocks.addAll(addedBlocks)
         routes.add(route)
+        for (block in addedBlocks) blockRoutes.getOrPut(block) { mutableSetOf() }.add(route)
```
Comment (shenriotpro) [inline core/src/main/kotlin/fr/sncf/osrd/stdcm/infra_exploration/InfraExplorer.kt:311]: nit: we may want to update the fun doc and add `blockRoutes`

**Review by shenriotpro — APPROVED** [review]

**Pull request merged**